runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -921,20 +925,434 @@ static void VMwrtbarEvaluator(TR::Node *node, TR::Register *srcReg, TR::Register
+   
+   TR::Register *fieldClassReg = NULL;
+   // Load fieldClass
+   if (isInstanceField)
+      {
+      fieldClassReg = cg->allocateRegister();
+      generateLoadJ9Class(node, fieldClassReg, sideEffectRegister, cg);
+      }
+   else if (!(node->getSymbolReference()->isUnresolved()))
+      {
+      fieldClassReg = cg->allocateRegister();
+      // During Non-AOT compilation the fieldClass has been populated inside the dataSnippet during compilation. 
+      // During AOT compilation the fieldClass must be loaded from the snippet. The fieldClass in an AOT body is invalid.
+      if (cg->comp()->compileRelocatableCode())
+         {
+         // Load FieldClass from snippet
+         generateTrg1MemInstruction(cg,TR::InstOpCode::Op_load, node, fieldClassReg,
+                           new (cg->trHeapMemory()) TR::MemoryReference(dataSnippetRegister, offsetof(J9JITWatchedStaticFieldData, fieldClass),
+                           TR::Compiler->om.sizeofReferenceAddress(), cg));
+         }
+      else
+         {
+         fieldClassReg = cg->allocateRegister();
+         J9Class * fieldClass = static_cast<TR::J9PPCWatchedStaticFieldSnippet *>(dataSnippet)->getFieldClass();
+         loadAddressConstant(cg, node, reinterpret_cast<uintptrj_t>(fieldClass), fieldClassReg);
+         }
+      }
+   else
+      {
+      // Unresolved 
+      if (isWrite)
+         {
+         fieldClassReg = cg->allocateRegister();
+         generateTrg1MemInstruction(cg,TR::InstOpCode::Op_load, node, fieldClassReg,
+               new (cg->trHeapMemory()) TR::MemoryReference(sideEffectRegister, fej9->getOffsetOfClassFromJavaLangClassField(), TR::Compiler->om.sizeofReferenceAddress(), cg));
+         }
+      else
+         {
+         fieldClassReg = sideEffectRegister;
+         }
+      } 
+
+   TR::MemoryReference *classFlagsMemRef = new (cg->trHeapMemory()) TR::MemoryReference(fieldClassReg, static_cast<uintptrj_t>(fej9->getOffsetOfClassFlags()),TR::Compiler->om.sizeofReferenceAddress(), cg);
+   
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
+   generateTrg1MemInstruction(cg,TR::InstOpCode::Op_load, node, scratchReg, classFlagsMemRef);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, scratchReg, scratchReg, cndReg, J9ClassHasWatchedFields);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, fieldReportLabel, cndReg);
+
+	generateReportOOL->swapInstructionListsWithCompilation();
+      
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, fieldReportLabel);
+   generateReportFieldAccessOutlinedInstructions(node, endLabel, dataSnippetRegister, isWrite, cg, sideEffectRegister, valueReg);
+      
+   generateReportOOL->swapInstructionListsWithCompilation();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, endLabel);
+   
+   cg->stopUsingRegister(cndReg);
+   cg->stopUsingRegister(scratchReg);
+   cg->stopUsingRegister(fieldClassReg);
```

Comment:
I think we have to be careful here now. With the change in 2c18083, it's possible that `fieldClassReg` will be `sideEffectRegister`. In those cases wouldn't it be incorrect to mark `fieldClassReg` as dead.

Reply:
Right. I think something like this would protect it from killing sideEffectReg 
https://github.com/eclipse/openj9/pull/6234/commits/b4ad4c8a8a26a50dcb38d8af78fe5c7b2c696292